components/support/sql/src/lazy.rs

```diff
@@ -0,0 +1,151 @@
+    // be contested and interrupt operations execute quickly.
+    interrupt_handle: Mutex<Option<Arc<SqlInterruptHandle>>>,
+}
+
+impl<CI: ConnectionInitializer> LazyDb<CI> {
+    /// Create a new LazyDb
+    ///
+    /// This does not open the connection and is non-blocking
+    pub fn new(path: &Path, open_flags: OpenFlags, connection_initializer: CI) -> Self {
+        Self {
+            path: path.to_owned(),
+            open_flags,
+            connection_initializer,
+            connection: Mutex::new(None),
+            interrupt_handle: Mutex::new(None),
+        }
+    }
+
+    /// Lock the database mutex and get a connection and interrupt scope.
+    ///
+    /// If the connection is closed, it will be opened.
+    ///
+    /// Calling `lock` again, or calling `close` while the mutex guard is still alive will cause a
+    /// deadlock.
+    pub fn lock(&self) -> Result<(MappedMutexGuard<'_, Connection>, SqlInterruptScope), Error> {
+        // Call get_conn first, then get_scope to ensure we acquire the locks in the correct order
+        let conn = self.get_conn()?;
+        let scope = self.get_scope(&conn)?;
+        Ok((conn, scope))
+    }
+
+    fn get_conn(&self) -> Result<MappedMutexGuard<'_, Connection>, Error> {
+        let mut guard = self.connection.lock();
+        // Open the database if it wasn't opened before.  Do this outside of the MutexGuard::map call to simplify the error handling
+        if guard.is_none() {
+            *guard = Some(open_database_with_flags(
+                &self.path,
+                self.open_flags,
+                &self.connection_initializer,
+            )?);
+        };
+        // Use MutexGuard::map to get a Connection rather than Option<Connection>.  The unwrap()
+        // call can't fail because of the previous code.
+        Ok(MutexGuard::map(guard, |conn_option| {
+            conn_option.as_mut().unwrap()
+        }))
+    }
+
+    fn get_scope(&self, conn: &Connection) -> Result<SqlInterruptScope, Error> {
+        let mut handle_guard = self.interrupt_handle.lock();
+        let result = match handle_guard.as_ref() {
+            Some(handle) => handle.begin_interrupt_scope(),
+            None => {
+                let handle = Arc::new(SqlInterruptHandle::new(conn));
+                register_interrupt(
+                    Arc::downgrade(&handle) as Weak<dyn AsRef<SqlInterruptHandle> + Send + Sync>
+                );
+                handle_guard.insert(handle).begin_interrupt_scope()
+            }
+        };
+        // If we see an Interrupted error when beginning the scope, it means that we're in shutdown
+        // mode.
+        result.map_err(|_| Error::Shutdown)
+    }
+
+    /// Close the database if it's open
+    ///
+    /// Pass interrupt=true to interrupt any in-progress queries before closing the database.
+    ///
+    /// Do not call `close` if you already have a lock on the database, as this will cause a
```

Comment:
Like mentioned above, this might sound too scary to consumers. There are still a few cases that calling `close()` is OK while a lock is held. For example, from a separate thread, they could call `interrup()` followed by `close()` as the former will interrupt in-progress queries and the lock will be released short after. Even if they don't interrupt, the close call will proceed whenever the lock is released. It'll block, but won't cause a deadlock.